src/proof.rs

```diff
@@ -597,16 +595,41 @@ impl<'a, 'b> ProofTreePrinterImpl<'a, 'b> {
         }
         process_block(&mut paren_stmt, &mut length_block);
 
+        self.write_word("(")?;
+        for s in &paren_stmt {
+            self.write_word(as_str(s.label()))?;
+        }
+        self.write_word(")")?;
+
+        let mut ess_stmt: Vec<StatementRef> = frame
+            .hypotheses
+            .iter()
+            .filter_map(|h| {
+                if let Hyp::Essential(addr, _) = h {
+                    Some(self.p.sset.statement(*addr))
+                } else {
+                    None
+                }
+            })
+            .collect();
+        ess_stmt.append(&mut paren_stmt);
+        paren_stmt = ess_stmt;
```

Comment:
Here we build the list of essential hypothesis, in their order of appearance in the database, i.e. from the scope pass, and prepend it to the list of statements.